Qq/Commands.lean

```diff
@@ -0,0 +1,103 @@
+import Qq.Macro
+import Lean
+
+namespace Qq
+open Lean Meta Elab Tactic
+
+private
+def mkHaveFVars (assignments : Array Expr) (body : Expr) : MetaM Expr := do
+  let ctx ← getLCtx
+  let (_, res) ← ctx.foldrM (init := (assignments, body)) fun decl (assignments, body) => do
+    let value := assignments.back!
+    let result := .letE decl.userName decl.type value (body.abstract #[.fvar decl.fvarId]) false
+    return (assignments.pop, result)
+  return res
+
+/--
+`by_elabq` is the Qq analogue to `by_elab`
+Executes arbitrary `TermElabM` code in place of a term.
+Moreover, all variables available in the local context are accessible as
+quoted expressions, and the return value is Q-annotated as well.
+Example:
+```
+def f (x : Prop) [Decidable x] : Int :=
+  by_elabq
+    Lean.logInfo x
+    Lean.logInfo x.ty
+    return q(if $x then 2 else 3)
+```
+See also: `run_tacq`.
+-/
+scoped
+elab "by_elabq" e:doSeq : term <= expectedType => do
+  let lctx ← getLCtx
+  let levelNames ← Term.getLevelNames
+  let (quotedCtx, assignments, quotedGoal) ← liftMetaM <| StateT.run'
+        (s := {mayPostpone := False}) do
+    let (quotedCtx, assignments) ← Impl.quoteLCtx lctx levelNames
+    let quotedGoal : Q(Type) ←
+      if expectedType.isMVar then pure q(TermElabM Expr)
+      else
+        let expectedTypeQ : Q(Expr) ← Qq.Impl.quoteExpr expectedType
+        pure <| q(TermElabM (Quoted $expectedTypeQ))
+    return (quotedCtx, assignments, quotedGoal)
+  let codeExpr : Expr ← withLCtx quotedCtx #[] do
+    let body ← Term.elabTermAndSynthesize (← `(do $e)) quotedGoal
+    mkHaveFVars assignments body
+  let code ← unsafe evalExpr (TermElabM Expr) q(TermElabM Expr) codeExpr
+  code
+
+/--
+`run_tacq` is a Qq-analogue to `run_tac`
+Executes an arbitrary `TacticM` code. Moreover, the local
+context can be directly accessed as quoted expressions.
+Optionally, user can also save the annotated goal using
+syntax `run_tacq $g =>`
+Example:
+```
+example (a b : Nat) (h : a = b) : True := by
+  run_tacq goal =>
+    let p : Q(Prop) := q($a = $b)
+    let t ← Lean.Meta.inferType h
+    Lean.logInfo p
+    Lean.logInfo <| toString (← Lean.Meta.isDefEq t p)
+    Lean.logInfo <| toString (← Lean.Meta.isDefEq h.ty p)
+    Lean.logInfo goal
+    Lean.logInfo goal.ty
+  trivial
+```
+See also: `by_elabq`.
+-/
+scoped
+syntax "run_tacq" (ident "=>")? doSeq : tactic
+
+elab_rules : tactic
+| `(tactic| run_tacq $[$gi:ident =>]? $seq:doSeq) => do
```

Comment:
I have to admit that syntax matching is still a bit an alchemy for me -- first I tried to do this with
```
elab "run_tacq" gi:(ident "=>")? seq:doSeq : tactic => do
```
which compiled but then only passed `none` to `gi`. With `elab_rules` it seems working.

Reply:
Turned out that the option syntax does not suffice in the end (it commits if the first is an ident). So now I have two cases for `elab_rules`.